diff -ur php-5.2.12.orig/configure.in php-5.2.12/configure.in
--- php-5.2.12.orig/configure.in	2009-12-16 12:45:18.000000000 +0000
+++ php-5.2.12/configure.in	2010-01-10 12:11:43.174983327 +0000
@@ -58,6 +58,8 @@
 AC_DEFUN([PHP_EXT_SRCDIR],[$abs_srcdir/ext/$1])dnl
 AC_DEFUN([PHP_ALWAYS_SHARED],[])dnl
 
+AC_DEFINE([_GNU_SOURCE], 1, [Define to enable GNU C Library extensions])
+
 dnl Setting up the PHP version based on the information above.
 dnl -------------------------------------------------------------------------
 
diff -ur php-5.2.12.orig/ext/interbase/interbase.c php-5.2.12/ext/interbase/interbase.c
--- php-5.2.12.orig/ext/interbase/interbase.c	2009-07-16 00:50:59.000000000 +0100
+++ php-5.2.12/ext/interbase/interbase.c	2010-01-10 12:11:43.176109071 +0000
@@ -24,7 +24,6 @@
 #include "config.h"
 #endif
 
-#define _GNU_SOURCE
 
 #include "php.h"
 
diff -ur php-5.2.12.orig/ext/pdo_firebird/firebird_driver.c php-5.2.12/ext/pdo_firebird/firebird_driver.c
--- php-5.2.12.orig/ext/pdo_firebird/firebird_driver.c	2009-04-18 19:56:58.000000000 +0100
+++ php-5.2.12/ext/pdo_firebird/firebird_driver.c	2010-01-10 12:11:43.177109181 +0000
@@ -22,7 +22,6 @@
 #include "config.h"
 #endif
 
-#define _GNU_SOURCE
 
 #include "php.h"
 #ifdef ZEND_ENGINE_2
diff -ur php-5.2.12.orig/ext/standard/file.c php-5.2.12/ext/standard/file.c
--- php-5.2.12.orig/ext/standard/file.c	2009-11-03 18:56:47.000000000 +0000
+++ php-5.2.12/ext/standard/file.c	2010-01-10 12:11:43.178108877 +0000
@@ -108,7 +108,6 @@
 
 #ifdef HAVE_FNMATCH
 #ifndef _GNU_SOURCE
-#define _GNU_SOURCE
 #endif
 #include <fnmatch.h>
 #endif
diff -ur php-5.2.12.orig/ext/zlib/zlib_fopen_wrapper.c php-5.2.12/ext/zlib/zlib_fopen_wrapper.c
--- php-5.2.12.orig/ext/zlib/zlib_fopen_wrapper.c	2009-01-20 15:43:06.000000000 +0000
+++ php-5.2.12/ext/zlib/zlib_fopen_wrapper.c	2010-01-10 12:11:43.180109009 +0000
@@ -19,8 +19,6 @@
 
 /* $Id: zlib_fopen_wrapper.c 274035 2009-01-20 15:43:06Z felipe $ */
 
-#define _GNU_SOURCE
-
 #include "php.h"
 #include "php_zlib.h"
 #include "fopen_wrappers.h"
diff -ur php-5.2.12.orig/main/php.h php-5.2.12/main/php.h
--- php-5.2.12.orig/main/php.h	2008-12-31 11:17:49.000000000 +0000
+++ php-5.2.12/main/php.h	2010-01-10 12:11:43.180109009 +0000
@@ -30,6 +30,7 @@
 #define PHP_HAVE_STREAMS
 #define YYDEBUG 0
 
+#include "php_config.h"
 #include "php_version.h"
 #include "zend.h"
 #include "zend_qsort.h"
diff -ur php-5.2.12.orig/main/streams/cast.c php-5.2.12/main/streams/cast.c
--- php-5.2.12.orig/main/streams/cast.c	2009-06-03 09:59:19.000000000 +0100
+++ php-5.2.12/main/streams/cast.c	2010-01-10 12:11:43.181109389 +0000
@@ -18,7 +18,6 @@
 
 /* $Id: cast.c 281581 2009-06-03 08:59:19Z kalle $ */
 
-#define _GNU_SOURCE
 #include "php.h"
 #include "php_globals.h"
 #include "php_network.h"
diff -ur php-5.2.12.orig/main/streams/memory.c php-5.2.12/main/streams/memory.c
--- php-5.2.12.orig/main/streams/memory.c	2009-09-17 03:45:25.000000000 +0100
+++ php-5.2.12/main/streams/memory.c	2010-01-10 12:11:43.182109190 +0000
@@ -18,7 +18,6 @@
 
 /* $Id: memory.c 288393 2009-09-17 02:45:25Z srinatar $ */
 
-#define _GNU_SOURCE
 #include "php.h"
 
 PHPAPI int php_url_decode(char *str, int len);
diff -ur php-5.2.12.orig/main/streams/streams.c php-5.2.12/main/streams/streams.c
--- php-5.2.12.orig/main/streams/streams.c	2009-05-17 15:59:24.000000000 +0100
+++ php-5.2.12/main/streams/streams.c	2010-01-10 12:11:43.184108025 +0000
@@ -21,7 +21,6 @@
 
 /* $Id: streams.c 280679 2009-05-17 14:59:24Z lbarnaud $ */
 
-#define _GNU_SOURCE
 #include "php.h"
 #include "php_globals.h"
 #include "php_network.h"
diff -ur php-5.2.12.orig/Zend/zend_language_parser.c php-5.2.12/Zend/zend_language_parser.c
--- php-5.2.12.orig/Zend/zend_language_parser.c	2009-12-16 12:48:14.000000000 +0000
+++ php-5.2.12/Zend/zend_language_parser.c	2010-01-10 12:11:43.187108409 +0000
@@ -356,6 +356,8 @@
 #include "zend_API.h"
 #include "zend_constants.h"
 
+#include <string.h>
+
 
 #define YYERROR_VERBOSE
 #define YYSTYPE znode
